### ISSUE TITLE
Need to setup essential modules

### ISSUE THREAD

**Issue by jvalleroy** [freedombox/Plinth#417]
Here's the error that I get when I try to run "plinth --setup" during freedombox-setup (by patching data/usr/lib/freedombox/setup.d/86_plinth).
```
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/psutil/_common.py", line 110, in wrapper
    return cache[key]
KeyError: (('/proc',), frozenset())

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/psutil/_pslinux.py", line 202, in <module>
    scputimes = set_scputimes_ntuple("/proc")
  File "/usr/lib/python3/dist-packages/psutil/_common.py", line 112, in wrapper
    ret = cache[key] = fun(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/psutil/_pslinux.py", line 184, in set_scputimes_ntuple
    with open_binary('%s/stat' % procfs_path) as f:
  File "/usr/lib/python3/dist-packages/psutil/_pslinux.py", line 138, in open_binary
    return open(fname, "rb", **kwargs)
FileNotFoundError: [Errno 2] No such file or directory: '/proc/stat'
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/psutil/_common.py", line 110, in wrapper
    return cache[key]
KeyError: (('/proc',), frozenset())

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/psutil/__init__.py", line 1435, in <module>
    _last_cpu_times = cpu_times()
  File "/usr/lib/python3/dist-packages/psutil/__init__.py", line 1429, in cpu_times
    return _psplatform.cpu_times()
  File "/usr/lib/python3/dist-packages/psutil/_pslinux.py", line 301, in cpu_times
    set_scputimes_ntuple(procfs_path)
  File "/usr/lib/python3/dist-packages/psutil/_common.py", line 112, in wrapper
    ret = cache[key] = fun(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/psutil/_pslinux.py", line 184, in set_scputimes_ntuple
    with open_binary('%s/stat' % procfs_path) as f:
  File "/usr/lib/python3/dist-packages/psutil/_pslinux.py", line 138, in open_binary
    return open(fname, "rb", **kwargs)
FileNotFoundError: [Errno 2] No such file or directory: '/proc/stat'
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/psutil/_common.py", line 110, in wrapper
    return cache[key]
KeyError: (('/proc',), frozenset())

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/psutil/__init__.py", line 1442, in <module>
    _last_per_cpu_times = cpu_times(percpu=True)
  File "/usr/lib/python3/dist-packages/psutil/__init__.py", line 1431, in cpu_times
    return _psplatform.per_cpu_times()
  File "/usr/lib/python3/dist-packages/psutil/_pslinux.py", line 314, in per_cpu_times
    set_scputimes_ntuple(procfs_path)
  File "/usr/lib/python3/dist-packages/psutil/_common.py", line 112, in wrapper
    ret = cache[key] = fun(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/psutil/_pslinux.py", line 184, in set_scputimes_ntuple
    with open_binary('%s/stat' % procfs_path) as f:
  File "/usr/lib/python3/dist-packages/psutil/_pslinux.py", line 138, in open_binary
    return open(fname, "rb", **kwargs)
FileNotFoundError: [Errno 2] No such file or directory: '/proc/stat'
[2016-02-29 00:05:44,290] plinth.__main__ INFO     Configured Django with applications - ['bootstrapform', 'django.contrib.auth', 'django.contrib.contenttypes', 'django.contrib.messages', 'stronghold', 'plinth', 'plinth.modules.apps', 'plinth.modules.avahi', 'plinth.modules.config', 'plinth.modules.datetime', 'plinth.modules.deluge', 'plinth.modules.diagnostics', 'plinth.modules.dynamicdns', 'plinth.modules.firewall', 'plinth.modules.first_boot', 'plinth.modules.help', 'plinth.modules.ikiwiki', 'plinth.modules.letsencrypt', 'plinth.modules.monkeysphere', 'plinth.modules.mumble', 'plinth.modules.names', 'plinth.modules.networks', 'plinth.modules.openvpn', 'plinth.modules.owncloud', 'plinth.modules.pagekite', 'plinth.modules.power', 'plinth.modules.privoxy', 'plinth.modules.quassel', 'plinth.modules.repro', 'plinth.modules.roundcube', 'plinth.modules.shaarli', 'plinth.modules.system', 'plinth.modules.tor', 'plinth.modules.transmission', 'plinth.modules.upgrades', 'plinth.modules.users', 'plinth.modules.xmpp']
[2016-02-29 00:05:44,291] plinth.__main__ INFO     Creating or adding new tables to data file
Operations to perform:
  Apply all migrations: auth, plinth, contenttypes
Running migrations:
  Rendering model states... DONE
  Applying contenttypes.0001_initial... OK
  Applying contenttypes.0002_remove_content_type_name... OK
  Applying auth.0001_initial... OK
  Applying auth.0002_alter_permission_name_max_length... OK
  Applying auth.0003_alter_user_email_max_length... OK
  Applying auth.0004_alter_user_username_opts... OK
  Applying auth.0005_alter_user_last_login_null... OK
  Applying auth.0006_require_contenttypes_0002... OK
  Applying auth.0007_alter_validators_add_error_messages... OK
  Applying plinth.0001_initial... OK
  Applying plinth.0002_modulestore...[2016-02-29 00:05:45,780] plinth.__main__ INFO     Configuration loaded from file - /etc/plinth/plinth.config
[2016-02-29 00:05:45,781] plinth.__main__ INFO     Script prefix - /plinth
[2016-02-29 00:05:45,781] plinth.module_loader INFO     Importing plinth.modules.apps
[2016-02-29 00:05:45,781] plinth.module_loader INFO     Importing plinth.modules.avahi
[2016-02-29 00:05:45,783] plinth.module_loader INFO     Importing plinth.modules.config
[2016-02-29 00:05:45,784] plinth.module_loader INFO     Importing plinth.modules.datetime
[2016-02-29 00:05:45,785] plinth.module_loader INFO     Importing plinth.modules.deluge
[2016-02-29 00:05:45,787] plinth.module_loader INFO     Importing plinth.modules.diagnostics
[2016-02-29 00:05:45,789] plinth.module_loader INFO     Importing plinth.modules.dynamicdns
[2016-02-29 00:05:45,792] plinth.module_loader INFO     Importing plinth.modules.firewall
[2016-02-29 00:05:45,793] plinth.module_loader INFO     Importing plinth.modules.first_boot
[2016-02-29 00:05:45,803] plinth.module_loader INFO     Importing plinth.modules.help
[2016-02-29 00:05:45,804] plinth.module_loader INFO     Importing plinth.modules.ikiwiki
[2016-02-29 00:05:45,806] plinth.module_loader INFO     Importing plinth.modules.letsencrypt
[2016-02-29 00:05:45,807] plinth.module_loader INFO     Importing plinth.modules.monkeysphere
[2016-02-29 00:05:45,808] plinth.module_loader INFO     Importing plinth.modules.mumble
[2016-02-29 00:05:45,812] plinth.module_loader INFO     Importing plinth.modules.names
[2016-02-29 00:05:45,813] plinth.module_loader INFO     Importing plinth.modules.networks
[2016-02-29 00:05:45,816] plinth.module_loader INFO     Importing plinth.modules.openvpn
[2016-02-29 00:05:45,818] plinth.module_loader INFO     Importing plinth.modules.owncloud
[2016-02-29 00:05:45,819] plinth.module_loader INFO     Importing plinth.modules.pagekite
[2016-02-29 00:05:45,822] plinth.module_loader INFO     Importing plinth.modules.power
[2016-02-29 00:05:45,823] plinth.module_loader INFO     Importing plinth.modules.privoxy
[2016-02-29 00:05:45,824] plinth.module_loader INFO     Importing plinth.modules.quassel
[2016-02-29 00:05:45,826] plinth.module_loader INFO     Importing plinth.modules.repro
[2016-02-29 00:05:45,827] plinth.module_loader INFO     Importing plinth.modules.roundcube
[2016-02-29 00:05:45,828] plinth.module_loader INFO     Importing plinth.modules.shaarli
[2016-02-29 00:05:45,829] plinth.module_loader INFO     Importing plinth.modules.system
[2016-02-29 00:05:45,830] plinth.module_loader INFO     Importing plinth.modules.tor
[2016-02-29 00:05:45,831] plinth.module_loader INFO     Importing plinth.modules.transmission
[2016-02-29 00:05:45,833] plinth.module_loader INFO     Importing plinth.modules.upgrades
[2016-02-29 00:05:45,834] plinth.module_loader INFO     Importing plinth.modules.users
[2016-02-29 00:05:45,838] plinth.module_loader INFO     Importing plinth.modules.xmpp
Failed to get unit file state for ejabberd.service: Unknown error 1178337856
Failed to get unit file state for repro.service: Unknown error 1700663872
Failed to get unit file state for quasselcore.service: Unknown error 378203584
Failed to get unit file state for transmission-daemon.service: Unknown error 1796043328
[2016-02-29 00:05:46,129] plinth.actions INFO     Executing command - ['sudo', '-n', '/usr/share/plinth/actions/pagekite', 'is-disabled']
Failed to connect to bus: No such file or directory
[2016-02-29 00:05:46,472] plinth.actions INFO     Executing command - ['sudo', '-n', '/usr/share/plinth/actions/pagekite', 'get-kite']
Failed to get unit file state for openvpn@freedombox.service: Unknown error 1753353664
Failed to get unit file state for mumble-server.service: Unknown error 1303264704
Failed to get unit file state for tor.service: Unknown error 1679020608
Failed to get unit file state for tor.service: Unknown error 1792486848
[2016-02-29 00:05:47,091] plinth.actions INFO     Executing command - ['sudo', '-n', '/usr/share/plinth/actions/tor', 'get-hs']
Failed to get unit file state for tor.service: Unknown error 874353216
ntp.service is not a native service, redirecting to systemd-sysv-install
Executing /lib/systemd/systemd-sysv-install is-enabled ntp
[2016-02-29 00:05:47,955] plinth.actions INFO     Executing command - ['/usr/share/plinth/actions/owncloud-setup', 'status']
Failed to get unit file state for privoxy.service: Unknown error 2055135808
[2016-02-29 00:05:47,981] plinth.setup   INFO     Running setup for all modules, essential - True
[2016-02-29 00:05:47,983] plinth.setup   INFO     Module does not require setup - apps
[2016-02-29 00:05:48,018] plinth.setup   INFO     Module does not require setup - help
[2016-02-29 00:05:48,051] plinth.setup   INFO     Module does not require setup - system
[2016-02-29 00:05:48,091] plinth.setup   INFO     Running module setup - avahi
[2016-02-29 00:05:48,091] plinth.setup   INFO     Running install for module - avahi, packages - ['avahi-daemon']
[2016-02-29 00:05:48,106] plinth.setup   ERROR    Error running setup - PackageException(error_string="Could not connect: No such file or directory", error_details="None")
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/plinth/package.py", line 89, in install
    self._do_install()
  File "/usr/lib/python3/dist-packages/plinth/package.py", line 103, in _do_install
    False, None, self.progress_callback, self)
GLib.GError: g-io-error-quark: Could not connect: No such file or directory (1)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/plinth/setup.py", line 77, in run
    self.module.setup(self, old_version=current_version)
  File "/usr/lib/python3/dist-packages/plinth/modules/avahi/__init__.py", line 65, in setup
    helper.install(['avahi-daemon'])
  File "/usr/lib/python3/dist-packages/plinth/setup.py", line 100, in install
    transaction.install()
  File "/usr/lib/python3/dist-packages/plinth/package.py", line 91, in install
    raise PackageException(exception.message) from exception
plinth.package.PackageException: PackageException(error_string="Could not connect: No such file or directory", error_details="None")
[2016-02-29 00:05:48,107] plinth.__main__ ERROR    Error running setup - PackageException(error_string="Could not connect: No such file or directory", error_details="None")
 OK
```

**Comment by jvalleroy** [comment]
Just for comparison, here's the error I get if I run plinth --setup during firstrun:
```
+ plinth --setup
[2016-02-29 00:44:35,283] plinth.__main__ INFO     Configured Django with applications - ['bootstrapform', 'django.contrib.auth', 'django.contrib.contenttypes', 'django.contrib.messages', 'stronghold', 'plinth', 'plinth.modules.apps', 'plinth.modules.avahi', 'plinth.modules.config', 'plinth.modules.datetime', 'plinth.modules.deluge', 'plinth.modules.diagnostics', 'plinth.modules.dynamicdns', 'plinth.modules.firewall', 'plinth.modules.first_boot', 'plinth.modules.help', 'plinth.modules.ikiwiki', 'plinth.modules.letsencrypt', 'plinth.modules.monkeysphere', 'plinth.modules.mumble', 'plinth.modules.names', 'plinth.modules.networks', 'plinth.modules.openvpn', 'plinth.modules.owncloud', 'plinth.modules.pagekite', 'plinth.modules.power', 'plinth.modules.privoxy', 'plinth.modules.quassel', 'plinth.modules.repro', 'plinth.modules.roundcube', 'plinth.modules.shaarli', 'plinth.modules.system', 'plinth.modules.tor', 'plinth.modules.transmission', 'plinth.modules.upgrades', 'plinth.modules.users', 'plinth.modules.xmpp']
[2016-02-29 00:44:35,284] plinth.__main__ INFO     Creating or adding new tables to data file
[2016-02-29 00:44:35,374] plinth.__main__ INFO     Configuration loaded from file - /etc/plinth/plinth.config
[2016-02-29 00:44:35,374] plinth.__main__ INFO     Script prefix - /plinth
[2016-02-29 00:44:35,375] plinth.module_loader INFO     Importing plinth.modules.apps
[2016-02-29 00:44:35,375] plinth.module_loader INFO     Importing plinth.modules.avahi
[2016-02-29 00:44:35,376] plinth.module_loader INFO     Importing plinth.modules.config
[2016-02-29 00:44:35,376] plinth.module_loader INFO     Importing plinth.modules.datetime
[2016-02-29 00:44:35,377] plinth.module_loader INFO     Importing plinth.modules.deluge
[2016-02-29 00:44:35,377] plinth.module_loader INFO     Importing plinth.modules.diagnostics
[2016-02-29 00:44:35,378] plinth.module_loader INFO     Importing plinth.modules.dynamicdns
[2016-02-29 00:44:35,389] plinth.module_loader INFO     Importing plinth.modules.firewall
[2016-02-29 00:44:35,390] plinth.module_loader INFO     Importing plinth.modules.first_boot
[2016-02-29 00:44:35,394] plinth.module_loader INFO     Importing plinth.modules.help
[2016-02-29 00:44:35,395] plinth.module_loader INFO     Importing plinth.modules.ikiwiki
[2016-02-29 00:44:35,396] plinth.module_loader INFO     Importing plinth.modules.letsencrypt
[2016-02-29 00:44:35,396] plinth.module_loader INFO     Importing plinth.modules.monkeysphere
[2016-02-29 00:44:35,396] plinth.module_loader INFO     Importing plinth.modules.mumble
[2016-02-29 00:44:35,418] plinth.module_loader INFO     Importing plinth.modules.names
[2016-02-29 00:44:35,429] plinth.module_loader INFO     Importing plinth.modules.networks
[2016-02-29 00:44:35,430] plinth.module_loader INFO     Importing plinth.modules.openvpn
[2016-02-29 00:44:35,431] plinth.module_loader INFO     Importing plinth.modules.owncloud
[2016-02-29 00:44:35,432] plinth.module_loader INFO     Importing plinth.modules.pagekite
[2016-02-29 00:44:35,433] plinth.module_loader INFO     Importing plinth.modules.power
[2016-02-29 00:44:35,433] plinth.module_loader INFO     Importing plinth.modules.privoxy
[2016-02-29 00:44:35,434] plinth.module_loader INFO     Importing plinth.modules.quassel
[2016-02-29 00:44:35,434] plinth.module_loader INFO     Importing plinth.modules.repro
[2016-02-29 00:44:35,435] plinth.module_loader INFO     Importing plinth.modules.roundcube
[2016-02-29 00:44:35,436] plinth.module_loader INFO     Importing plinth.modules.shaarli
[2016-02-29 00:44:35,436] plinth.module_loader INFO     Importing plinth.modules.system
[2016-02-29 00:44:35,436] plinth.module_loader INFO     Importing plinth.modules.tor
[2016-02-29 00:44:35,437] plinth.module_loader INFO     Importing plinth.modules.transmission
[2016-02-29 00:44:35,438] plinth.module_loader INFO     Importing plinth.modules.upgrades
[2016-02-29 00:44:35,438] plinth.module_loader INFO     Importing plinth.modules.users
[2016-02-29 00:44:35,451] plinth.module_loader INFO     Importing plinth.modules.xmpp
[2016-02-29 00:44:35,454] plinth.actions INFO     Executing command - ['sudo', '-n', '/usr/share/plinth/actions/pagekite', 'is-disabled']
[2016-02-29 00:44:35,824] plinth.actions INFO     Executing command - ['sudo', '-n', '/usr/share/plinth/actions/pagekite', 'get-kite']
Failed to get unit file state for ejabberd.service: No such file or directory
Failed to get unit file state for mumble-server.service: No such file or directory
Failed to get unit file state for transmission-daemon.service: No such file or directory
Failed to get unit file state for repro.service: No such file or directory
Failed to get unit file state for privoxy.service: No such file or directory
Failed to get unit file state for tor.service: No such file or directory
Failed to get unit file state for tor.service: No such file or directory
[2016-02-29 00:44:36,334] plinth.actions INFO     Executing command - ['sudo', '-n', '/usr/share/plinth/actions/tor', 'get-hs']
Failed to get unit file state for tor.service: No such file or directory
Failed to get unit file state for quasselcore.service: No such file or directory
ntp.service is not a native service, redirecting to systemd-sysv-install
Executing /lib/systemd/systemd-sysv-install is-enabled ntp
Failed to get unit file state for openvpn@freedombox.service: No such file or directory
[2016-02-29 00:44:36,512] plinth.actions INFO     Executing command - ['/usr/share/plinth/actions/owncloud-setup', 'status']
[2016-02-29 00:44:36,514] plinth.setup   INFO     Running setup for all modules, essential - True
[2016-02-29 00:44:36,515] plinth.setup   INFO     Module does not require setup - system
[2016-02-29 00:44:36,523] plinth.setup   INFO     Module does not require setup - users
[2016-02-29 00:44:36,527] plinth.setup   INFO     Module does not require setup - apps
[2016-02-29 00:44:36,533] plinth.setup   INFO     Module does not require setup - names
[2016-02-29 00:44:36,538] plinth.setup   INFO     Running module setup - firewall
[2016-02-29 00:44:36,538] plinth.setup   INFO     Running install for module - firewall, packages - ['firewalld']
[2016-02-29 00:44:36,766] plinth.setup   ERROR    Error running setup - PackageException(error_string="Cannot refresh cache whilst offline", error_details="None")
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/plinth/package.py", line 89, in install
    self._do_install()
  File "/usr/lib/python3/dist-packages/plinth/package.py", line 103, in _do_install
    False, None, self.progress_callback, self)
GLib.GError: pk-client-error-quark: Cannot refresh cache whilst offline (257)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/plinth/setup.py", line 77, in run
    self.module.setup(self, old_version=current_version)
  File "/usr/lib/python3/dist-packages/plinth/modules/firewall/__init__.py", line 60, in setup
    helper.install(['firewalld'])
  File "/usr/lib/python3/dist-packages/plinth/setup.py", line 100, in install
    transaction.install()
  File "/usr/lib/python3/dist-packages/plinth/package.py", line 91, in install
    raise PackageException(exception.message) from exception
plinth.package.PackageException: PackageException(error_string="Cannot refresh cache whilst offline", error_details="None")
[2016-02-29 00:44:36,769] plinth.__main__ ERROR    Error running setup - PackageException(error_string="Cannot refresh cache whilst offline", error_details="None")
Operations to perform:
  Apply all migrations: contenttypes, auth, plinth
Running migrations:
  No migrations to apply.
```

**Comment by SunilMohanAdapa** [comment]
To make PackageKit not use network manager for detection of network, in /etc/PackageKit/PackageKit.conf we can write:
> UseNetworkManager=false

But we would like run `plinth --setup` during setup and not firstboot, so this is not enough.

**Comment by SunilMohanAdapa** [comment]
I found that running
```
service dbus start
```
Before
```
plinth --start
```
seems to take things further.

**Comment by jvalleroy** [comment]
Looks like we also need to mount proc. I put this in freedombox-customize, before setup is run:
```
mount proc $rootdir/proc -t proc
```
With this I was able to get the setup to run, but only on one system (a FreedomBox). On my desktop, it still fails with the GLib.GError.

**Comment by SunilMohanAdapa** [comment]
On 03/04/2016 05:34 AM, James Valleroy wrote:
> Looks like we also need to mount proc. I put this in
> freedombox-customize, before setup is run:
> 
> |mount proc $rootdir/proc -t proc |
> 
> With this I was able to get the setup to run, but only on one system (a
> FreedomBox). On my desktop, it still fails with the GLib.GError.
> 

I have actually tested inside a chroot after mounting all required
filesystems.  I didn't do a full check (hence no patch yet).

mount /dev -t devfs -obind $rootdir/dev
mount /dev/pts -t devpts -obind $rootdir/dev/pts
mount /proc -t proc -obind $rootdir/proc
mount /sys -t sysfs -obind $rootdir/sys

This is what is available on a running system and this is what
vmdeboostrap uses for setting up grub/uefi.  I wrongly assumed
vmdebootstrap will do this for customize script also.  I checked again
the source and realized.

-- 
Sunil
